### PR TITLE
#394: Use Union[List[Column], List[str]] for Select

### LINKED ISSUE #394: Use covariant `Sequence` for `select()`
Having an array `a=["col1", "col2"]` the expression `df.select(a)` we get the following mypy error:

```
file.py:10: error: Argument 1 to "select" of "DataFrame" has incompatible type "List[str]"; expected "List[Union[Column, str]]"  [arg-type]
file.py:10: note: "List" is invariant -- see http://mypy.readthedocs.io/en/latest/common_issues.html#variance
file:10: note: Consider using "Sequence" instead, which is covariant
Found 1 error in 1 file (checked 1 source file)
```

### PR DESCRIPTION
**Pull request by qudade**
Passing a `List[str]` to select raises a mypy warning, similar for `List[Column]`. We change the type from `List[Union[Column, str]]` to `Union[List[Column], List[str]]`.

Fixes #394 .

### CODE REVIEW

**Comment by zero323** [comment]
That looks much better. Would you mind adding [data tests](https://github.com/zero323/pyspark-stubs/tree/master/test-data/unit) for this?

**Comment by qudade** [comment]
@zero323 sure, I'll try to add some tests next week.

Is there some information how to run the tests? (simply calling `pytest` didn't seem to do the trick)

**Comment by zero323** [comment]
> @zero323 sure, I'll try to add some tests next week.

Thanks.

> Is there some information how to run the tests? (simply calling `pytest` didn't seem to do the trick)

You'll have to add annotations directory to your `MYPYPATH` and set `MYPY_TEST_PREFIX`, i.e.

https://github.com/zero323/pyspark-stubs/blob/c3e949caec8673d781fcdb151f2a3f06a10de4ac/.travis.yml#L17

**Comment by qudade** [comment]
@zero323 Tests have been added. Code was rebased vs master and squashed.

Tests are passing (except for #397)

File: test-data/unit/sql-dataframe.test [hunk 1]
```diff
@@ -21,6 +21,77 @@ df.sample(withReplacement=False) # E: No overload variant of "sample" of "DataFr
...
+data = [('Alice', 1)]
+df = spark.createDataFrame(data, schema="name str, age int")
+
+df.groupBy(["name", "age"])
+df.groupby(["name", "age"])
+df.groupBy([col("name"), col("age")])
+df.groupby([col("name"), col("age")])
+
+
+df.groupBy(["name", col("age")])  # E: Argument 1 to "groupBy" of "DataFrame" has incompatible type "List[object]"; expected "Union[List[Column], List[str]]"
```
Comment (zero323) [inline test-data/unit/sql-dataframe.test:55]: I guess this is an acceptable trade-off.

Comment (qudade) [reply test-data/unit/sql-dataframe.test:55]: I hope so. pyspark considers this mixture legal, but I have seen many `List[str]` or `List[Column]` and never such a hybrid.

It's a shame though that we can't get everything...

**Comment by zero323** [comment]
Merged to master, thank you.

**Pull request merged**